### DISCUSSION TITLE
🍬  `when`, `unless`, `ifElse` -- sweet, sweet sugar over `choose` causing problems with the VSCode extension

### DISCUSSION THREAD

**Discussion by Silverwolf90** [statelyai/xstate#2662] [Ideas]
I've got a set of helpers that wrap around `choose` which provide cleaner patterns for branching actions:

```ts
export const when = (cond, actions) =>
  choose([
    { cond, actions }
  ]);

export const ifElse = (cond, ifTrue, ifFalse) =>
  choose([
    { cond, actions: ifTrue },
    { actions: ifFalse }
  ]);

export const unless = (cond, actions) =>
  ifElse(cond, [], actions);
```

Examples:

```ts
when('someCondition', 'someAction'),
when('someCondition', ['someAction', 'anotherAction']),

unless('someCondition', 'someAction'),
unless('someCondition', ['someAction', 'anotherAction']),

ifElse('someCondition', 'someAction', 'yetAnotherAction'),
ifElse('someCondition', ['someAction', 'anotherAction'], 'yetAnotherAction'),
ifElse('someCondition', ['someAction', 'anotherAction'], ['yetAnotherAction', 'andYetAnother']),

```

Obviously these are pretty trivial to implement. So it seems like a clear case for something that belongs in "userland" -- which I would normally agree with. But since xstate is looking to become an ecosystem, patterns that can't take advantage of the ecosystem's tools are effectively dead in the water.

In the current state of the VSCode extension (keep in mind, it's at a super early stage) these helper functions come at a severe compromise to DX:

- Can't do "Go to definition"/cmd+click
- Guards/actions are flagged for not being used (even though they are)
- Not recognized by the inspector as built-in actions (shows up as  "anonymous")

[I had a brief discussion with @mattpocock in the #dev-experience discord channel about this](https://discord.com/channels/795785288994652170/808385662512267274/887690476461916160), and it sounds like it would be quite hard to parse these "userland" things correctly. So in my head I think "well maybe this belongs in xstate core then!" 😆

I do not consider these helpers to be opinionated, but merely sugar that enables readable, concise branching. `xstate` suffers enough from being "syntax heavy" and `choose` is very noisy for simple cases. If I had to choose 1 of these to make it into the core library, it would be `when` which we use extensively -- we have ~60 occurrences across our machines.

Thoughts?

**Comment by davidkpiano** [comment]
Wow, I love this actually.

In v5, we have `and(...)`, `or(...)`, and `not(...)` helpers for guards.

In SCXML, there is [`<if>`](https://www.w3.org/TR/scxml/#if), [`<elseif>`](https://www.w3.org/TR/scxml/#elseif), and [`<else>`](https://www.w3.org/TR/scxml/#else) for actions.

Since we obviously can't have an `if(...)` action creator, `when(...)` seems like an intuitive substitute.

In v5, `unless(...)` can also be represented succinctly as `when(not(...), ...)`, so `unless(...)` can stay in userland.

And then for now, `ifElse(...)` is just a little shorter than `choose([{ ... }, { ... }])`, so we'll see based on usage if that should be brought into core.

But I'd be glad to consider adding `when(...)` 👍

**Reply by Silverwolf90** [reply]
Great to hear! Totally agree that `when(not(...), ...)` works as a replacement for `unless`. And also agree that `ifElse` provides marginal benefit over `choose`.

**Reply by mattpocock** [reply]
I find the `ifElse` the most difficult to read!

These funcs are small enough in signature, especially 'when', that they could be a nice addition to core.

**Reply by davidkpiano** [reply]
@Silverwolf90 Want to make a PR for `when(...)`?

**Reply by Silverwolf90** [reply]
Yes will do!
